### PR TITLE
feat_: batch all telemetry data and send request every 10 seconds

### LINKED ISSUE #5240: bug: had noticed in multiple instances that enabling telemetry is causing message loss
# Problem
When telemetry is enabled, noticing logs from gossipSub layer indicating that application buffer is full and hence some of the received messages are not getting delivered and hence dropped.
This can lead to message loss. 

In 3 different running desktop instances, we had noticed this. 

# Analysis

On going through the message processing code in waku and status-go, had noticed that go-routine that processes incoming messages also pushes data to telemetry server(which is a blocking HTTP call). This causes delay in message processing in-turn causing buffered channel to get full. When this happens, gossipsub drops messages and records below log. 

`  3582 INFO [05-27|17:22:10.582|go.uber.org/zap/sugar.go:168]                                                               Can't deliver message to subscription for topic /waku/2/rs/16/32; subscriber too slow`

# Suggested Solution

In order to avoid processing delay in received messages(which seems to be only due to telemetry for now), we should modify telemetry Client to use a channel and a go-routine to feed off of the channel. Note that enough buffer can be kept in this channel to handle slowness of telemetry response. 
This would ensure that message processing is not delayed and buffer doesn't become full. 

Also, another optimization that can be done in telemetry is to batch submit telemetry information to the telemetry server. This would ensure that amount of HTTP requests to telemetry server can be reduced.  

# Acceptance Criteria

Have a test to send a lot of messages and ensure that message processing routine doesn't get blocked. 

# Notes

We need to make sure to add comments in the message processing method to ensure no blocking calls are there and offload any network related interactions into separate go-routines. 

# Future Steps

None

### PR DESCRIPTION
**Pull request by adklempner**
Instead of making an http request for every individual telemetric, push each to a channel and periodically batch in a single request.

A description to understand introduced changes without reading the code.

Important changes:
- [ ] Depends on https://github.com/status-im/telemetry/pull/16

Closes #5240

### CODE REVIEW

File: wakuv2/telemetry.go [hunk 1]
```diff
@@ -50,5 +50,7 @@ func (c *BandwidthTelemetryClient) PushProtocolStats(relayStats metrics.Stats, s
 	_, err := c.httpClient.Post(url, "application/json", bytes.NewBuffer(body))
 	if err != nil {
 		c.logger.Error("Error sending message to telemetry server", zap.Error(err))
+	} else {
+		c.logger.Debug("Successfully pushed protocol stats to telemetry server")
```
Comment (richard-ramos) [inline wakuv2/telemetry.go]: Might be too noisy since this is the expected result. Consider not logging this

File: telemetry/client.go [hunk 2]
```diff
@@ -96,5 +233,7 @@ func (c *Client) UpdateEnvelopeProcessingError(shhMessage *types.Message, proces
 	_, err := c.httpClient.Post(url, "application/json", bytes.NewBuffer(body))
 	if err != nil {
 		c.logger.Error("Error sending envelope update to telemetry server", zap.Error(err))
+	} else {
+		c.logger.Debug("Successfully pushed envelope processing error to telemetry server", zap.String("hash", types.EncodeHex(shhMessage.Hash)))
```
Comment (richard-ramos) [inline telemetry/client.go]: Might be too noisy due to it being the expected result. Consider not logging this

File: telemetry/client.go [hunk 3]
```diff
@@ -10,70 +10,207 @@ import (
...
+			case receivedEnvelope := <-c.receivedEnvelopeCh:
+				c.processAndPushTelemetry(receivedEnvelope)
+			case sentEnvelope := <-c.sentEnvelopeCh:
+				c.processAndPushTelemetry(sentEnvelope)
+			}
+		}
+	}()
+}
+
+func (c *Client) Start() {
```
Comment (richard-ramos) [inline telemetry/client.go]: Should we pass a context to the Start() function?
We could use it for killing the go routine once logout happens, i.e.:

```
select {
...
case <-ctx.Done():
    return
...
}
```

File: telemetry/client.go [hunk 4]
```diff
@@ -10,70 +10,207 @@ import (
...
+		receivedMessagesCh: make(chan wakuv2.ReceivedMessages),
+		receivedEnvelopeCh: make(chan *v2protocol.Envelope),
+		sentEnvelopeCh:     make(chan wakuv2.SentEnvelope),
+		telemetryCh:        make(chan TelemetryRequest),
+		nextId:             0,
+		sendPeriod:         10 * time.Second,
+	}
+}
+
+func (c *Client) CollectAndProcessTelemetry() {
```
Comment (richard-ramos) [inline telemetry/client.go]: Perhaps we should accept a context here too to exit the goroutine on logout and avoid having a dangling goroutine?  (The messenger should have a `m.ctx` context you could use)

File: telemetry/client.go [hunk 5]
```diff
@@ -10,70 +10,207 @@ import (
...
+		return
 	}
+
+	c.nextId++
+	c.telemetryCh <- telemetryRequest
 }
 
-func (c *Client) PushReceivedMessages(filter transport.Filter, sshMessage *types.Message, messages []*v1protocol.StatusMessage) {
+func (c *Client) pushTelemetryRequest(request []TelemetryRequest) {
+	c.logger.Debug("Pushing telemetry data", zap.Any("request", request))
```
Comment (richard-ramos) [inline telemetry/client.go]: Consider not logging this unless it's strictly necessary

**Review by richard-ramos — APPROVED** [review]
Nice PR!
Just left few comments and questions for your consideration

**Review by endulab — APPROVED** [review]

File: telemetry/client.go (hunk 2 above)
Comment (adklempner) [reply telemetry/client.go]: removed in latest commit

File: telemetry/client.go (hunk 3 above)
Comment (adklempner) [reply telemetry/client.go]: Added in latest commit

File: telemetry/client.go (hunk 4 above)
Comment (adklempner) [reply telemetry/client.go]: Added in latest commit

File: telemetry/client.go (hunk 5 above)
Comment (adklempner) [reply telemetry/client.go]: removed in latest commit

File: wakuv2/telemetry.go (hunk 1 above)
Comment (adklempner) [reply wakuv2/telemetry.go]: removed in latest commit

**Pull request merged**